### ISSUE TITLE
[security][3.3] Backport smtplib fix for TLS stripping vulnerability, CVE-2016-0772

### ISSUE THREAD

**Issue by vstinner** [python/cpython#74770]
BPO | [30585](https://bugs.python.org/issue30585)
--- | :---
Nosy | @birkenfeld, @vstinner, @ned-deily
PRs | <li>python/cpython#225</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = 'https://github.com/birkenfeld'
closed_at = <Date 2017-07-19.00:46:42.571>
created_at = <Date 2017-06-07.08:36:32.909>
labels = ['type-security']
title = '[security][3.3] Backport smtplib fix for TLS stripping vulnerability, CVE-2016-0772'
updated_at = <Date 2017-07-19.00:46:42.569>
user = 'https://github.com/vstinner'
```

bugs.python.org fields:
```python
activity = <Date 2017-07-19.00:46:42.569>
actor = 'ned.deily'
assignee = 'georg.brandl'
closed = True
closed_date = <Date 2017-07-19.00:46:42.571>
closer = 'ned.deily'
components = []
creation = <Date 2017-06-07.08:36:32.909>
creator = 'vstinner'
dependencies = []
files = []
hgrepos = []
issue_num = 30585
keywords = []
message_count = 4.0
messages = ['295319', '295320', '298620', '298621']
nosy_count = 3.0
nosy_names = ['georg.brandl', 'vstinner', 'ned.deily']
pr_nums = ['225']
priority = None
resolution = 'fixed'
stage = 'resolved'
status = 'closed'
superseder = None
type = 'security'
url = 'https://bugs.python.org/issue30585'
versions = ['Python 3.3']
```

</p></details>

**Comment by vstinner** [comment]
Attached pull request backports a fix for this security vulnerability:
http://python-security.readthedocs.io/vuln/cve-2016-0772_smtplib_tls_stripping.html

"A vulnerability in smtplib allowing MITM attacker to perform a startTLS stripping attack. smtplib does not seem to raise an exception when the remote end (SMTP server) is capable of negotiating starttls but fails to respond with 220 (ok) to an explicit call of SMTP.starttls(). This may allow a malicious MITM to perform a startTLS stripping attack if the client code does not explicitly check the response code for startTLS."

Reported by: Tin (Team Oststrom)

Python 2.7, 3.4, 3.5 and 3.6 are already safe.

**Comment by vstinner** [comment]
PEP-398: Python 3.3 Release Schedule

Python 3.3 branch end of support is expected to be at 2017-09-29, in 4 months.

**Comment by ned-deily** [comment]
New changeset 3625f7fd11679ecb390ffa58ef36d487acc8159b by Ned Deily (Victor Stinner) in branch '3.3':
[3.3] bpo-30585: [security] raise an error when STARTTLS fails (#225)
https://github.com/python/cpython/commit/3625f7fd11679ecb390ffa58ef36d487acc8159b

**Comment by ned-deily** [comment]
Merged for release in 3.3.7rc1
